### ISSUE TITLE
get/any errors out on unset paths

### ISSUE THREAD

**Issue by hiiamboris** [red/red#4910]
**Describe the bug**
```
>> o: object [unset a: 'a]
== make object! [
    a: unset
]
>> get/any 'o/a
*** Script Error: o/a has no value
*** Where: get
*** Stack:  
```

**Expected behavior**

Returns unset.

**Platform version**
```
Red 0.6.4 for Windows built 12-May-2021/6:12:03+03:00  commit #cebd5b9
```

**Comment by hiiamboris** [comment]
More info here: https://github.com/red/red/issues/4550#issuecomment-671937637

**Comment by dockimbel** [comment]
Defining the semantics of `get/any` for paths is not that straightforward. Consider path with more than two elements:
```
get/any 'a/b/c/d/e
== unset
```
Which one of the path words is actually `unset`? We can't know. What if `a/b/c` part is actually returning `none`, then `/d` is erroring out. But what if `d` is unset, which error should it return?

**Comment by hiiamboris** [comment]
This ticket concerns only with the final value, that is `a/b/c/d` should be valid.
I expect error still be thrown if any of the intermediate path words are not indexable as paths (although error message can be more specific and at least tell you which token in the path led to the error)

**Comment by dockimbel** [comment]
> This ticket concerns only with the final value, that is a/b/c/d should be valid.

That is then even more problematic. In `get/any 'a/b/c/d/e`, if `c` is unset, you get an error, but if `e` only is unset, you get no error... inconsistent and confusing.

**Comment by hiiamboris** [comment]
On the contrary I think it would be consistent with how get-paths work.

**Comment by dockimbel** [comment]
Get-path exhibits also this inconsistent behavior in the processing of `unset` depending on the position in the path. This looks to me like something that needs fixing rather than supporting.
```
>> p: object [a: object [t: object [h: none set/any 'h ()]]]       ; h == unset
>> :p/a/t/h                 ; no error
>> set/any 'p/a ()          ; a == unset
>> :p/a/t/h
*** Script Error: path :p/a/t/h is not valid for unset! type
*** Where: catch
*** Stack:  
```

**Comment by hiiamboris** [comment]
Why is it inconsistent?
- `1/none/()/x` is an invalid path, and it fails on all scalars in the path
- while `block/block/unset` is a valid path that yields an unset value, and this is where `/any` should by design muffle the `unset`-related error, because we're ready to handle it

**Comment by greggirwin** [comment]
I'm with @hiiamboris here, in that the evaluation _of_ an `unset!` (for something further)  is what triggers the error. Evaluting _to_ `unset!` is OK.

**Comment by dockimbel** [comment]
My concern was not theoretical but practical.  I think the following will be confusing for most users:
```
>> get/any 'a/b/c/d/e
*** Script Error: path a/b/c/d/e is not valid for unset! type
```

**Comment by hiiamboris** [comment]
Because the message is confusing. It should tell the part of the path that evaluated to unset. Rebol error is clearer IIRC.

**Comment by dockimbel** [comment]
Fixing the error message would help indeed. I'll look into it.

**Comment by dockimbel** [comment]
I have improved the error message. Still, I think the errors will be confusing to most users:
```
>> a: context [b: context [c: 123]]
>> get/any 'a/b/c
== 123

>> get/any 'a/z/c
** Script error: cannot access z in path a/z/c
** Where: get
** Near: get/any 'a/z/c

>> get/any 'a/b/z
(unset)
```
I think that explaining to users why `/any` is effective on the last word only but not on intermediary ones will be challenging.
Maybe I am lowering the bar too much here, I'm not sure. @greggirwin what do you think?

**Comment by greggirwin** [comment]
Compared to the original issue, this is a big improvement. Could the error message be:
```
Script error: cannot access z/c in path a/z/c because z is unset
```
or
```
Script error: cannot access c in path a/z/c because z is unset
```
?

A lot of things are challenging, so we just have to try to make them better incrementally, until the cost of making the better is too high; hoping the cost comes down in the future.

**Comment by dockimbel** [comment]
> Could the error message be:
```
Script error: z is unset in path a/z/c
```
would be the cheapest option, as it requires only changing the error message string. The others you proposed requires allocating and constructing a new sub-path just for error message. If such error is nested in a loop and wrapped in a `try`, it will generate additional pressure on the GC.

**Comment by greggirwin** [comment]
That's fine. I wouldn't be terribly concerned about error cases like this and GC impact. Yes, we want to be efficient, but most edge cases are not as important as usability.

**Comment by dockimbel** [comment]
Extra unit tests are pending.

**Comment by hiiamboris** [comment]
```
>> a: object [unset b: 'b c: 1]
== make object! [
    b: unset
    c: 1
]

>> get/any 'a/b/1
*** Script Error: b is unset in path a/b/1
*** Where: get
*** Stack:

>> get/any 'a/c/1
*** Script Error: path a/c/1 is not valid for integer! type
*** Where: get
*** Stack:
```
@dockimbel While you're at it, I suggest patching error message for all non-indexable types, not just unset.

**Comment by greggirwin** [comment]
1) `b is an integer! in path a/b/1`
2) `b is not an indexable* type in path a/b/1`

*2 raises the question of what we call these types. Indexable, keyed, compound, ...accessor-ized?

1 is probably good enough for now. It at least says what value had a problem, and doesn't go into details we may need to change later, depending on what we end up calling things.

**Comment by hiiamboris** [comment]
I'm for (1) as it tells the type, which is a valuable piece of info. Or `b is an integer! in path a/b/1 and does not support path access` then no need to invent a name (which is hard considering functions and files also support path access with different meaning than indexing).

**Comment by dockimbel** [comment]
`b is an integer!...`
`b is a tuple!...`

Error messages can't handle `a/an` dynamically. We need a formulation that stays the same for all type names.

**Comment by hiiamboris** [comment]
`a/1/1` - good point. I also thought about duplicate word like `a/x/b/x`... Paths that Gregg proposes would solve the ambiguities, e.g. `a/1 is of integer! type and does not support path access` (or `a/x` or `a/x/b/x`...). As for GC pressure, we could worry about that later when we have slices. Also, how big is the actual pressure coming from an extra path, compared to creation of error object (and a hash for it)?

**Comment by greggirwin** [comment]
`a/1 is a type (integer!) that does not support path access` ?

Though it seems that in this day and age we should be able to handle a/an. ;^)

But look at all the fun details this has brought out, which I call progress.

**Comment by dockimbel** [comment]
`a/1 is a datatype (integer!)...` This formulation sounds confusing to me, `a/1` is a path! not an integer!.

**Comment by hiiamboris** [comment]
I consider error messages of high priority as they save (or waste) time of every Reducer. Until 1.0 there'll be a lot of them... ;)

**Comment by greggirwin** [comment]
I agree that semantic and advanced edge cases are not as important as some other things, and also that error messages are important. Making them informative and friendly helps to fight complexity in user's minds. Since there is no easy way around the natural language formulations (which we should consider for future designs related to templated output, including `format`) we'll have to live with a less natural construction.

1)
`a/b/1`
>> b is a datatype (integer!) in path a/b/1 and does not support path access

`a/1/1`
>> 1 is a datatype (integer!) in path a/1/1 and does not support path access

2)
`a/b/1`
>> b is of type integer! in path a/b/1 and does not support path access

`a/1/1`
>> 1 is of type integer! in path a/1/1 and does not support path access

3)
`a/b/1`
>> b is a/an integer! in path a/b/1 and does not support path access

`a/1/1`
>> 1 is a/an integer! in path a/1/1 and does not support path access

4)
???

**Comment by dander** [comment]
4.
`a/b/1`

> a/b in path a/b/1 is of type integer! and does not support path access

`a/1/1`

>  a/1 in path a/1/1 is of type integer! and does not support path access

**Comment by dockimbel** [comment]
> I consider error messages of high priority as they save (or waste) time of every Reducer. Until 1.0 there'll be a lot of them... ;)

I don't see how that way of prioritizing can lead to a 1.0. Even simply outputting error codes instead of messages would be acceptable to me in order to reach a fully featured and stable 1.0.

**Comment by hiiamboris** [comment]
Is it going to take more than an hour to provide a good error message?

**Comment by dockimbel** [comment]
We spent already many man-hours on this ticket. The point is not how much time/resources is spent on one specific ticket, but how many "details" do we want to fix/improve instead of working on the big features on which depends the future of Red. We are still in alpha stage, not 1.x. The rules are not the same, the goals are not the same. Satisfying end users by polishing every possible detail should not be part of the work during alpha stage. Priority and importance are two different qualifiers. Many features and tickets are important, but only a very few are high-priority.

In a nutshell, focusing on fixing details is holding back the whole development of Red. We need to change that if we want a 1.0 ever to see the light.

**Comment by hiiamboris** [comment]
Totally agree. This issue was never a big thing. I'd more like to see #4854 done for instance.
Of course, since your cache is already filled with this issue, would be wasteful to leave it half-solved ;)

As for road to the stars (1.0 :), IMO there's a shorter way to that.. e.g. if you concentrated on designs, and left R/S grunt work to others... PRs are a powerful thing, and community participation if nourished and guided could get us there much faster. One man is just one man, with only 24 hours a day.

**Comment by GiuseppeChillemi** [comment]
> In a nutshell, focusing on fixing details is holding back the whole development of Red. We need to change that if we want a 1.0 ever to see the light. Polishing details is the goal of 0.9.x version. We don't even have a Red language documentation which at this point should be a much higher-priority than fixing any non-crashing/corrupting bug.

I agree, I can sacrifice precise error messages until 0.9 or even past 1.0. The most important thing is having Red complete, after that we can concentrate on the details.

**Comment by hiiamboris** [comment]
Like this https://github.com/red/red/issues/4532 ?

**Comment by dockimbel** [comment]
@hiiamboris I wish it would be that simple. It cannot be designed only from an abstract perspective, often implementation limits the design or orient it in a different way. They are quite intricated. Though, I agree that I should write down more design descriptions and guideline rules that contributors could leverage.
